packages/react/src/extend-theme.ts

```diff
@@ -21,45 +25,143 @@ type DeepThemeExtension<BaseTheme, ThemeType> = {
     : CloneKey<ThemeType, Key>
 }
 
-export type ThemeOverride = Partial<ChakraTheme> &
-  DeepThemeExtension<Theme, ChakraTheme>
+export type ThemeOverride<BaseTheme = Theme> = Partial<ChakraTheme> &
+  DeepThemeExtension<BaseTheme, ChakraTheme> &
+  Dict
+
+export type ThemeExtension<Override extends ThemeOverride = ThemeOverride> = (
+  themeOverride: Override,
+) => Override
+
+export type BaseThemeWithExtensions<
+  BaseTheme extends ChakraTheme,
+  Extensions extends readonly [...any]
+> = Extensions extends [infer L, ...infer R]
+  ? L extends (...args: any[]) => any
+    ? ReturnType<L> & BaseThemeWithExtensions<BaseTheme, R>
+    : L & BaseThemeWithExtensions<BaseTheme, R>
+  : ThemeOverride<BaseTheme>
```

Comment:
This type is TS 4.1 - we need to find a way to express this with TS3.8 or consider bumping the min. TS version.

Reply:
Nevermind - just saw that we support TS ^4.1.3